### PR TITLE
Allow disabling of user/org creation in the HASH workspace

### PR DESCRIPTION
**Pull request by benwerner01**
## 🌟 What is the purpose of this PR?

This PR makes it possible to disable user and org creation in the HASH App, by introducing two new properties on the HASH instance entity type:
1. `userSelfRegistrationIsEnabled` - whether or not a user can self register (i.e. sign-up) with HASH.
1. `userRegistrationByInviteIsEnabled` - whether or not an existing user is able to register another user by inviting them to an org. Note this is not enforced as we don't have an org invite flow.
1. `orgSelfRegistrationIsEnabled` - whether or not a user can self-register an org (note this does not apply to instance admins). Note this is not enforced as we don't have a org registration flow for users.

When both (1) and (2) are `false`, user accounts can only be created by an instance administrator's action.

When only (1) is `false`, existing users will still be able to invite other users to a HASH instance (via -- when supported -- sharing an entity/type with a specific email address, or by adding them to an org).

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203301853897838/1203249971975780/f) _(internal)_

## 🔍 What does this change?

See commits.

## 📜 Does this require a change to the docs?

No.

## 🛡 What tests cover this?

None.

## ❓ How to test this?

1. Run HASH using `yarn dev`
3. Observe that the sign up button is being displayed
4. Login as any user
5. Use the `updateEntityWithMetadata` GQL query to set `userSelfRegistrationIsEnabled` to `false`
```GraphQL
mutation updateEntityWithMetadata($updatedProperties:JSONObject!) {
  updateEntityWithMetadata(
    entityId:"00000000-0000-0000-0000-000000000000%<hashInstanceUuid>",
    updatedProperties: $updatedProperties
  ) 
}
```

```JSON
{
 "updatedProperties": {
  "http://localhost:3000/@example-org/types/property-type/user-self-registration-is-enabled/": true,
  ...remainingProperties,
}
```
6. Sign out, and observe that the sign up button disappears from the UI.

### CODE REVIEW

File: packages/hash/api/src/graphql/resolvers/knowledge/hashInstance/hashInstance.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  QueryGetAllLatestEntitiesWithMetadataArgs,
+} from "../../../apiTypes.gen";
+import { LoggedInGraphQLContext } from "../../../context";
+
+export const hashInstanceEntity: ResolverFn<
+  Promise<EntityWithMetadata>,
+  {},
+  LoggedInGraphQLContext,
+  QueryGetAllLatestEntitiesWithMetadataArgs
+> = async (_, __, { dataSources }, ___) => {
```
Comment (CiaranMn) [inline packages/hash/api/src/graphql/resolvers/knowledge/hashInstance/hashInstance.ts:14]:
```suggestion
> = async (_, __, { dataSources }) => {
```

File: packages/hash/api/src/model/knowledge/hashInstance.model.ts [hunk 2]
```diff
@@ -234,4 +242,16 @@ export default class extends EntityModel {
 
     await outgoingAdminLinkEntityModel.archive(graphApi, { actorId });
   }
+
+  isUserRegistrationEnabled(): boolean {
+    return (this.getProperties() as any)[
```
Comment (CiaranMn) [inline packages/hash/api/src/model/knowledge/hashInstance.model.ts:247]: Not a suggestion for this PR, but when cleaining up the model classes I would consider having `PropertyObject` to take a generic which can then be fed through from classes that extend `Entity` to properly type their `properties`.

**Review by CiaranMn — COMMENTED** [review]
Code looks good – I couldn't test this as it needs updating for name changes (e.g. `/signup` doesn't load because a mutation's name has been changed) – happy to do a quick test and approve once that's updated

File: packages/hash/api/src/model/knowledge/hashInstance.model.ts (hunk 2 above)
Comment (Alfred-Mountfield) [reply packages/hash/api/src/model/knowledge/hashInstance.model.ts:247]: I like the idea of properly typing them, but you can use `PropertyObject` here to type this a lot better (a `BaseUri` keyed JSON object)

Comment (benwerner01) [reply packages/hash/api/src/model/knowledge/hashInstance.model.ts]: `getProperties()` actually now returns `PropertyObject`, so the typing can be improved slightly:
https://github.com/hashintel/hash/blob/9facccde72bf80819227de06ea2f2c02b5e3bb75/packages/hash/api/src/model/knowledge/hashInstance.model.ts#L246-L256

File: packages/hash/shared/src/types.ts [hunk 3]
```diff
@@ -156,6 +156,20 @@ const systemPropertyTypes = {
...
+  userRegistrationByInviteIsEnabled: {
+    title: "User Registration By Invitation Is Enabled",
+    description:
+      "Whether or not a user is able to register another user by inviting them to an org.",
+  },
+  orgSelfRegistrationIsEnabled: {
+    title: "Org Self Registration Is Enabled",
+    description:
+      "Whether or not a user can self-register an org (note this does not apply to instance admins).",
+  },
```
Comment (benwerner01) [inline packages/hash/shared/src/types.ts:172]: Note to reviewers, these don't actually do anything yet because we don't currently have flows for "invite user to an org" or "register an org".

File: packages/hash/api/src/auth/index.ts [hunk 4]
```diff
@@ -55,6 +55,14 @@ const kratosAfterRegistrationHookHandler =
       try {
         const { emails } = traits;
 
+        const hashInstanceModel = await HashInstanceModel.getHashInstanceModel(
+          graphApi,
+        );
+
+        if (!hashInstanceModel.isUserSelfRegistrationEnabled()) {
```
Comment (thehabbos007) [inline packages/hash/api/src/auth/index.ts:62]: Should we also be checking for `userRegistrationByInviteIsEnabled`? Or is that registration handled elsewhere?

File: packages/hash/api/src/model/knowledge/user.model.ts [hunk 5]
```diff
@@ -154,6 +154,10 @@ export default class extends EntityModel {
       isInstanceAdmin = false,
     } = params;
 
+    const hashInstanceModel = await HashInstanceModel.getHashInstanceModel(
```
Comment (thehabbos007) [inline packages/hash/api/src/model/knowledge/user.model.ts:157]: Why was this moved outside if we don't need it for adding an admin? Do we need to conduct any test?

File: packages/hash/api/src/model/util.ts [hunk 6]
```diff
@@ -330,6 +330,9 @@ export type LinkEntityTypeCreatorParams = Omit<
   "entityTypeId"
 > & {
   linkEntityTypeId: VersionedUri;
+  title: string;
+  description?: string;
+  actorId: string;
```
Comment (thehabbos007) [inline packages/hash/api/src/model/util.ts:335]: Do these not exist on `EntityTypeCreatorParams`?

File: packages/hash/shared/src/types.ts [hunk 7]
```diff
@@ -311,6 +325,7 @@ type DataTypeDefinition = TypeDefinition & { dataTypeId: VersionedUri };
 
 type LinkEntityTypeDefinition = TypeDefinition & {
   linkEntityTypeId: VersionedUri;
+  description?: string;
```
Comment (thehabbos007) [inline packages/hash/shared/src/types.ts:328]: Same as before for `EntityTypeCreatorParams`. Does `TypeDefinition` not contain `description`?

**Review by thehabbos007 — COMMENTED** [review]
Thank you! I have some comments :)

File: packages/hash/api/src/auth/index.ts (hunk 4 above)
Comment (benwerner01) [reply packages/hash/api/src/auth/index.ts:62]: > Should we also be checking for userRegistrationByInviteIsEnabled?

No, that would have to be checked in some flow dedicated to invitations.

> Or is that registration handled elsewhere?

It's handled nowhere, because currently there is no org invite flow

File: packages/hash/api/src/model/knowledge/user.model.ts (hunk 5 above)
Comment (benwerner01) [reply packages/hash/api/src/model/knowledge/user.model.ts]: This wan't moved intentionally, reverted in https://github.com/hashintel/hash/pull/1489/commits/3b3308eea0904bfd51e50e7d79b864a44d213f3d

File: packages/hash/api/src/model/util.ts (hunk 6 above)
Comment (benwerner01) [reply packages/hash/api/src/model/util.ts]: `description` actually wasn't, but have added as part of https://github.com/hashintel/hash/pull/1489/commits/ae331c6d485b3adc957d258418d4f012f70ac308

File: packages/hash/shared/src/types.ts (hunk 7 above)
Comment (benwerner01) [reply packages/hash/shared/src/types.ts]: True, addressed in https://github.com/hashintel/hash/pull/1489/commits/ae331c6d485b3adc957d258418d4f012f70ac308

**Review by thehabbos007 — APPROVED** [review]
